### ISSUE TITLE
JobOptions.cs

### ISSUE THREAD

**Issue by RFBomb** [tjscience/RoboSharp#125]
@PCAssistSoftware, @JoE-Seeve,  @MarkECrum

I'm opening a new Issue Thread for JobOptions.CS discussion here. I'll be pasting all the current posts I'm aware of related to it in the next few minutes so we have a single location to collaborate on this addition.

**Comment by RFBomb** [comment]
JobOptions.CS as proposed by Joe over in [#109 ](https://github.com/tjscience/RoboSharp/issues/109#issuecomment-970652699)

```
using System.Text;

namespace RoboSharp
{
    public class JobOptions
    {
        internal const string JOBNAME_LOAD = "/JOB:{0} ";
        internal const string JOBNAME_SAVE = "/SAVE:{0} ";
        internal const string QUIT_AFTER_PROCESSING = "/QUIT ";
        internal const string NO_SOURCE_DIRECTORY = "/NOSD ";
        internal const string NO_DESTINATION_DIRECTORY = "/NODD ";
        internal const string INCLUDE_FILES = "/IF ";

        private string jobNameLoad = string.Empty;
        private string jobNameSave = string.Empty;

        /// <summary>
        /// Take parameters from the named job file (default file name is ""). [/JOB:jobname]
        /// </summary>
        public string JobNameLoad
        {
            get { return jobNameLoad; }
            set { jobNameLoad = value; }
        }

        /// <summary>
        /// Save parameters to the named job file (default file name is "") [/SAVE:jobname].
        /// </summary>
        public string JobNameSave
        {
            get { return jobNameSave; }
            set { jobNameSave = value; }
        }

        /// <summary>
        /// Quit after processing command line [/QUIT].
        /// </summary>
        public bool QuitAfterProcessing { get; set; }

        /// <summary>
        /// No source directory is specified [/NOSD].
        /// </summary>
        public bool NoSourceDirectory { get; set; }

        /// <summary>
        /// No destination directory is specified [/NODD].
        /// </summary>
        public bool NoDestinationDirectory { get; set; }

        /// <summary>
        /// No destination directory is specified [/NODD].
        /// </summary>
        public bool IncludeFiles { get; set; }

        internal string Parse()
        {
            var options = new StringBuilder();

            if (!JobNameLoad.IsNullOrWhiteSpace())
                options.Append(string.Format(JOBNAME_LOAD, JobNameLoad));

            if (!JobNameSave.IsNullOrWhiteSpace())
                options.Append(string.Format(JOBNAME_SAVE, JobNameSave));

            if (NoSourceDirectory)
                options.Append(NO_SOURCE_DIRECTORY);

            if (NoDestinationDirectory)
                options.Append(NO_DESTINATION_DIRECTORY);

            if (IncludeFiles)
                options.Append(INCLUDE_FILES);

            if (QuitAfterProcessing)
                options.Append(QUIT_AFTER_PROCESSING);

            return options.ToString();
        }
    }

}
```

**Comment by RFBomb** [comment]
Related Comment: https://github.com/tjscience/RoboSharp/pull/113#issuecomment-994898938

> I just took a quick second look at the JOB portion of that thread and did a quick read up on what the Job Options entails for robocopy. Nice info guide: https://adamtheautomator.com/robocopy/#Robocopy_Jobs
> 
> In summary, and correct me if I'm wrong, essentially the 'Job Options' allows for saving a robocopy command(s) to a text file so that it can be used later. What this allows is essentially building out the parameters into multiple files for later recall. Based on my testing, this **IS NOT** able to be used to multiple robocopy runs in succession. This is because is you specify 2 job files, each with a SOURCE or DESTINATION specified, then what you end up with is 2 sources and 2 destinations, which conflicts. 
> 
> What you CAN do is put all the store options like IncludedFiles / ExcludedFiles, etc into these job files, then when you call them up it will essentially concat all those options together as needed. 
> Ex:
> ```
> File 1 Source: C:\Source  /IF *.exe  /XF *.txt
> File 2 Destination: C:\Dest  /IF *.jpg 
> result:
> `robocopy C:\Source C:\Dest /IF *.exe *.jpg  /XF *.txt`
> 
> ```
> BUT if you have:
> 
> ```
> File 1 Source: C:\Source  Destination: C:\Dest /IF *.exe  /XF *.txt
> File 2 Source: C:\Source2 Destination: C:\Dest2  /IF *.jpg 
> result:
> ` Invalid Parameter #2 : "/JOB:JobTest"`
> ```
> Robocopy will report an invalid parameter due to specifying multiple sources/destinations. 
>  
> In summary, JobOptions cannot be used for multiple robocopy commands within this project currently. It may be possible to store a single command / options into job files then parse them though. This parsing would be quite an intensive little routine though. But in order for multiple jobs to run, it would still have to be implemented in a List format.
> 

I had written a parser for Job Options files that robocopy produces. It requires the remaining items to be implemented:

```
/// <summary>
        /// Parse some Job File
        /// </summary>
        /// <param name="filePath">File Extension should end in *.RCJ</param>
        /// <returns>New <see cref="RoboSharp.RoboCommand"/></returns>
        public static RoboSharp.RoboCommand ParseJobFile(string filePath) 
        {
            RoboSharp.RoboCommand cmd = new RoboSharp.RoboCommand();
            cmd.JobOptions.ParseJobFile(filePath);
            return cmd;
        }                


/// <summary>
        /// Parse some Job File
        /// </summary>
        /// <param name="filePath">File Extension should end in *.RCJ</param>
        /// <returns>New <see cref="RoboSharp.RoboCommand"/></returns>
        public void ParseJobFile(string filePath)
        {
            RoboSharp.RoboCommand cmd = this;
            if (File.Exists(filePath)
            {
                string[] lines = File.ReadAllLines(filePath));
                string l;
                string[] s;
                bool parsingIF = false;
                bool parsingXD = false;
                bool parsingXF = false;
                foreach (string line in lines)
                {
                    if (line.Trim().StartsWith("::"))
                    { /*comment line - ignore */
                        parsingIF = false;
                        parsingXD = false;
                        parsingXF = false;
                    }
                    //Source
                    else if (line.Trim().StartsWith("/SD"))
                    {
                        s = line.Split(':');
                        cmd.CopyOptions.Source = s[1];
                    }
                    //Destination
                    else if (line.Trim().StartsWith("/DD"))
                    {
                        s = line.Split(':');
                        cmd.CopyOptions.Destination = s[1];
                    }
                    //Include Files
                    else if (parsingIF || line.Trim().StartsWith("/IF"))
                    {
                        if (parsingIF)
                            cmd.CopyOptions.FileFilter += $"{line.RemoveJobComment()}";
                        else
                            parsingIF = true;
                    }
                    //Exclude Directories
                    else if (parsingXD || line.Trim().StartsWith("/XD"))
                    {
                        if (parsingXF)
                            cmd.SelectionOptions.ExcludeDirectories += $"{line.RemoveJobComment()}";
                        else
                            parsingXF = true;
                    }
                    //Exclude Files
                    else if (parsingXF || line.Trim().StartsWith("/XF"))
                    {
                        if (parsingXF)
                            cmd.SelectionOptions.ExcludeFiles += $"{line.RemoveJobComment()}";
                        else
                            parsingXF = true;
                    }

                    #region < Copy Options > 
                    else if (line.Trim().StartsWith("/PURGE"))
                        cmd.CopyOptions.Purge = true;
                    else if (line.Trim().StartsWith("/COPY:"))
                    { 
                        //Must parse
                    }
                    else if (line.Trim().StartsWith("/DCOPY:"))
                    {
                        //Must parse
                    }
                    #endregion

                    #region < Retry Options > 
                    else if (line.Trim().StartsWith("/R:"))
                    {
                        l = line.RemoveJobComment().split(':')[1];
                        cmd.RetryOptions.RetryCount = Convert.ToInt32(l);
                    }
                    else if (line.Trim().StartsWith("/W:"))
                    {
                        l = line.RemoveJobComment().split(':')[1];
                        cmd.RetryOptions.RetryWaitTime = Convert.ToInt32(l);
                    }
                    #endregion

                    #region < Logging Options > 
                    else if (line.Trim().StartsWith("/R:"))
                    {
                        //To Do
                    }
                    else if (line.Trim().StartsWith("/W:"))
                    {
                        
                    }
                    #endregion
                }
            }

        }
```

Note: these will have to be added to the StringExtensions.cs:

```
/// <summary>Trim the comment off of a line parsed from a RoboCopy.RCJ Job File</summary>
internal static string RemoveJobComment(this string str) => str.Trim().RemoveTrailingString("::").Trim();

        /// <summary>Find the start of the <paramref name="SearchTerm"/> in the string, and remove everything after it.</summary>
        /// <param name="SearchTerm"> String to search for.</param>
        /// <param name="str">String to parse.</param>
        /// <returns>
        /// Example 1: <br/>
        /// Input String: "TestString"  -- SearchTerm: "Str" <br/>
        /// Output: "Test" <para/>
        /// Example 2: <br/>
        /// Input String: "TestString"  -- SearchTerm: "X" <br/>
        /// Output: "TestString"  - Search Term was not found.
        /// </returns>
        public static string RemoveTrailingString(this string str, string SearchTerm)
        {
            if (str.IsNullOrEmpty()) return "";
            int i = str.LastIndexOf(SearchTerm);
            return (i >= 0) ? str.Remove(i) : str;
        }

```

**Comment by RFBomb** [comment]
Followup-comment: https://github.com/tjscience/RoboSharp/pull/113#issuecomment-994926377

> 
> @JoE-Seeve created a nice starting point for JobOptions.cs, I believe is mergable without issue. ~~It just needs to be properly documented with `<summary>` and `<see href="""/>` tags.~~ Hopefully that is something that @MarkECrum  and @JoE-Seeve can collaborate on to get implemented.

(Upon review, Joe already did the remarks!)

> 
> They should also include the parser I framed out above into JobOptions.cs in order to be able to load the options into a robocommand object, rather then relying on calling the file itself. This way, RoboSharp can still report what the options actually are. The parser can either run when they issue the 'start' command along with all the other parser stuff, but I think it may be better to run immediately to reflect the new options in the object, rather than at runtime. Either way, they will have to flesh out the parser and verify functionality.
> 
> To have the best of both worlds, I would create a new object and create a list of them for the JobOptions.cs to avoid parsing already-parsed jobs, then create a `List<JobFile>` property and a "AddJobFile" method to add to the list. of Something like:
> 

```
public class JobFile {

#region < Constructor >
private JobFile(){}

public JobFile(RoboCommand cmd, string filePath)
{ 
    RoboCmd = cmd;
     FilePath = filePath; 
}


public JobFile(RoboCommand cmd, string filePath, bool ParseImmediately)
{ 
    RoboCmd = cmd;
     FilePath = filePath; 
    if ( ParseImmediately ) this.Parse();
}
#endregion

#region < Properties >

/// <summary>The RoboCommand object this JobFile is assigned to </summary>
public RoboCommand RoboCmd {get; private set;}

/// <summary>Log if this JobFile has been parsed yet </summary>
public bool HasBeenParsed {get; private set;}

/// <summary>FilePath of the Job File </summary>
public string FilePath {get;set;}

#endregion

#region < Methods >

/// <summary> Reset the HasBeenParsed flag </summary>
public void ResetHasBeenParsed() => HasBeenParsed = false;

//Put the Parser method I wrote above here - Make sure to update HasBeenParsed=true once complete !

#endregion


}
```

**Comment by JoE-Seeve** [comment]
My intention to integrate JobOptions.cs was actually the following points:
- Saving of copy jobs defined in a RoboSharp App in a format that others can understand or process (e.g. RoboCopy or RoboSharp Apps).
- A RoboSharp app can read these files back in and use them again with a parser (reading and parsing is fine in my test, but not quite finished yet). What is meant here is a single copy job that can be processed.
- I also thought that you can read in several job files and run RoboSharp commands with them and work in parallel. So I actually thought that you can do the copying work in parallel within an app without starting the same app multiple times (which is definitely possible).
- But I didn't think of merging the various job files and using the result to do a copy job. RoboCopy allows some merging via intermediate files (or something like that).

Here is an example of how I am currently using it (save via a RoboSharp command, load via its own parser):
![grafik](https://user-images.githubusercontent.com/92209494/146230735-fc5a0d2e-6f91-4b53-bb73-1c7af4e8f886.png)

**Comment by RFBomb** [comment]
the merging I refer to is what RoboCopy does on its own when you specify multiple job files. 

It sounds more like what you are saying is specify a job file, parse it into a robocommand object (which is what I agree with). 
But, since robocopy allows for the merge, we would also have to accomodate that functionality since this is a library. 

Best not to re-invent the wheel, since robocopy can already spit out job files, we should just parse the job files robocopy generates, which is where I came up with my proposed parser method.

Reading a job file into a RoboCommand would allow the end-developer to use/show it however they want.

**Comment by JoE-Seeve** [comment]
@RFBomb Yes, that was the actual basic idea that you can at least process a single file again first (but I wasn't there yet).

Is there a way to control multiple instances of RoboCopy / RoboSharp Command via a list?

What's the best way to get ahead? Should I open a pull request? I don't have any experience with that, but I will definitely manage it somehow (I hope).

**Comment by RFBomb** [comment]
Create a fork of the repo, then create a new branch based off the the main branch here, that way your branch is starting off using the latest accepted changes here. Create the file and make adjustments as needed.

Once you are satisfied, you can perform a 'commit', then create a pull request. Much of this process is greatly simplified by Visual Studio if you have the Git / github extension installed. If using VS-Code, I use GitHub Desktop to perform commits and PRs

**Comment by RFBomb** [comment]
As for controlling multiple instances via a list, that currently is up to the application the uses this library, its not built in. 

What I do for my application is create the list, and loop through it in 2 stages. 
Stage 1 turns the List-Only switch on, and runs the list. 

```
List<Task> tasks = new List<Task>();
 foreach (RoboCopy cmd in RoboCopyCommands)
            {
                cmd.CopyOptions.Threads = 3;
                tasks.Add(cmd.CountFilesAsync()); //This is used in my Wrapper to turn ListOnly on, run RoboCommand.Start(), then turn ListOnly off
            }
 await Task.WhenAll(tasks); // Wait for all scanning operations  to finish asyncronously
```

Stage 2 runs through the loop but the copy operation is performed 1 robocommand at a time. Although, you could run multiple at once using the code described above.

for cancelling/pausing:

```
foreach (RoboCopy cmd in RoboCopyCommands)
                    {
                        cmd.Stop();
                    }
```

**Comment by JoE-Seeve** [comment]
Yes, that's exactly what I had in mind with a list of RoboCopyCommands. If JobOptions is integrated, predefined copy options can be started via job files. 

If you get the parameters directly from a job file, then it would be helpful to transfer them directly. But then GenerateParameters in RoboComand.cs  would have to be adapted and a direct input of the parameters would have to be possible.

**Comment by RFBomb** [comment]
Thats where the parse method comes in. You would parse all the JobFile objects, then run the rest of the GenerateParameters to generate the actual call. The idea here is that parsing the JobFiles would add all those options into the RoboCommand object.

**Comment by RFBomb** [comment]
Just to clarify here (for future reference), this thread has to do with Saving & Loading jobs to/from disk. This thread is not for running multiple robocommands in succession/simeultaneously 

#126 will be a good thread to use for running multiple robocommands.

**Comment by MarkECrum** [comment]
Hey guys... Sorry I have been coding.... I am very willing to work on this section. However, I once was listed as a Microsoft Professional Developer (between 1998 and < 2010.....  I can code, But it was a  long time ago and I am still learning C# and the Visual Studio UI. So.... What needs done for solid implementation???
